test_runner/regress/test_replication_start_subxid_overflow.py

```diff
@@ -0,0 +1,217 @@
+import threading
+import time
+
+from fixtures.neon_fixtures import NeonEnv, wait_for_last_flush_lsn, wait_replica_caughtup
+
+
+def initdb(cur):
+    cur.execute("create table t(pk serial primary key, payload integer)")
+    cur.execute(
+        """create or replace function create_subxacts(n integer) returns void as $$
+              declare
+                 i integer;
+              begin
+                 for i in 1..n loop
+                    begin
+                       insert into t (payload) values (0);
+                    exception
+                       when others then
+                          raise exception 'caught something';
+                    end;
+                 end loop;
+              end; $$ language plpgsql"""
+    )
+
+
+def create_subxacts(env, endpoint, n_subxacts, barrier):
+    con = endpoint.connect()
+    cur = con.cursor()
+    cur.execute("create schema background")
+    cur.execute("set search_path=background")
+    initdb(cur)
+    cur.execute("begin")
+    cur.execute(f"select create_subxacts({n_subxacts})")
+    wait_for_last_flush_lsn(env, endpoint, env.initial_tenant, env.initial_timeline)
+    barrier.wait()
+    time.sleep(5)  # wait for replica start
+    cur.execute("commit")
+
+
+# This test replica startup in case of large number of active subtransactions at primary node.
+# Number of known transaction xids is limited and if we do not limit number of transactions
+# restored from CLOG, then it will cause crash of replica.
```

Comment:
As it stands, this test passed even without the code changes from this PR.

Reply:
> The key thing not mentioned in this comment is that we now wait for the transactions in the primary to finish. Right?

Sorry, May be I do not completely undertand the question. "Large number of active substransactions at primary" IMHO obviously mean that at the moment of replica startup there were active transactions.

The only think which is change is that we can not explicitly commit it AFTER replica startup. Because it will not start until this transaction is active. We have to launch thread for it.

Reply:
> As it stands, this test passed even without the code changes from this PR.

No, it doesn't

Reply:
Tested this again on my laptop. Code is with fresh checkout from `main`, and the tests from this PR:
```
[gw0] [ 33%] PASSED test_runner/regress/test_replication_start_subxid_overflow.py::test_replication_start_subxid_overflow[debug-pg16] 
[gw1] [ 66%] FAILED test_runner/regress/test_replication_start_subxid_overflow.py::test_replication_start_subxid_overflow2[debug-pg16] 
[gw0] [100%] FAILED test_runner/regress/test_replication_start_subxid_overflow.py::test_replication_start_subxid_overflow3[debug-pg16] 
```

I claimed earlier that `test_replication_start_subxid_overflow2' also passes on 'master', but that seems to be false. I must've messed up something in my testing. But `test_replication_start_subxid_overflow` _does_ pass.

Reply:
It is not passed at my environment.
Last week I spent time mostly to make tests os this PR not only stable pass but also stable fail:)
I mean failure of main branch and catch "too much knowns xacts" error in `test_replication_start_subxid_overflow3`.

I think we have consensus on this PR: limitations and possible drawbacks of this approach fro restoring running xaxts from CLOG:
- We can not prevent "too much known factors" error
- If there was subxid overflow and primary is not active than startup of replica will be blocked.
Probability of both problems is very small: no so much people are using subxacts and even if sub transaction are used, their number very unlikely exceed few thousands.

Yes, it is possible to make this test so deterministic but by making them even more complex: instead of two threads (as now), have three threads  master thread, replica thread and orchestration thread. But don't you think that it its overkill and instead of fixing real problem with replication, we spent a lot of efforts just for test developement.

I rather prefer to incorporate your CSN approach rather than try precisely checks all drawbacks of the current approach.